### PR TITLE
Expose proximity boosting

### LINKED ISSUE #33382: Expose proximity boosting
In addition to static features (ctr, pagerank, url length) it is common to also want to boost by recency or geo-distance, which are going to perform differently on every request depending on the current time or on the location of the user that runs the query.

Lucene just added support for efficient [boosting by recency](https://issues.apache.org/jira/browse/LUCENE-8340) and plans to add something similar for [boosting by geo-distance](https://issues.apache.org/jira/browse/LUCENE-8482).

How should we expose it in Elasticsearch?

Relates #27588
cc @polyfractal @mayya-sharipova

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Expose DistanceFeatureQuery for geo, date_nanos an date types

Closes #33382

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/docbldesx

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: nit: can be private

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: Why are you catching IAE here? I don't think we usually do that but just bubble up exceptions instead. Maybe it has advantages I'm missing though.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: Can you check START_ARRAY explicitely here and otherwise throw an error if we encounter an unexpected token?

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: nit: private

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: same as above, please check the expected token type and throw error if anything unexpected shows up in the token stream.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: Since all arguments are required, add a null-check please.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: this could move to an inner class if we added them for Origin and Pivot.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: nit: space between ")?"

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: nit: space between "(String)pivot"

File: server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java]: nit: spaces

File: server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java]: nit: spaces

**Review by cbuescher — CHANGES_REQUESTED** [review]
@mayya-sharipova this looks like a really cool and useful feature. I left a couple of comments, most of them minor. I think the biggest suggestion I'd have would be trying to avoid to have untyped (Object) origin and pivot fields and instead create some very simple inner class for holding this information in either one of the allowed types (long, String, Date). This way we should be able to control which values the user can set and prevent accidental misuses early.
I was also wondering if it makes sense to add a few integration tests (something extending ESIntegTestCase) to check if the scoring works as advertised (e.g. boosts are applied correctly, pivot works as expected etc...) but then again I guess this is already covered in the underlying Lucene queries so maybe its not that useful, but wanted to bring it up as a suggestion.
Other than that great feature, looking forward to using it in 7.1.

File: docs/reference/query-dsl/distance-feature-query.asciidoc
Comment (mayya-sharipova) [reply docs/reference/query-dsl/distance-feature-query.asciidoc]: @cbuescher I have copied this phrase from [rank_feature query](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/query-dsl-rank-feature-query.html) which is using the same optimizations, and we also have a [blog post](https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand) on this.  This blog post has a link to Lucene benchmarks, but looks like  adding this link to these benchmarks would be excessive here.

File: docs/reference/query-dsl/special-queries.asciidoc
Comment (jpountz) [reply docs/reference/query-dsl/special-queries.asciidoc]: should it mention dates?

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: this isn't correct with `date_nanos`.

**Review by jpountz — COMMENTED** [review]
This looks good overall. Two thoughts on the PR:
 - I'm wondering that we should only expose it on `date(_nanos)` and `geo_point` fields for now. Lucene exposes it only longs mostly because it doesn't have dedicated fields for dates. As-is I think it feels a bit inconsistent to support this feature on longs but not integers or doubles. Since it's also likely less useful than recency boosting I'm wondering that we should leave it out for now?
 - When I opened the issue I wondered whether it would make more sense to have one query for all data types or one query per data type. Looking at the PR I'm wondering that having one query per data type might make things cleaner. For instance it's a pity that some validation only occurs at the shard level while we could do it at the coordinating node level if we knew whether the query would apply to a geo point or a date field.

**Comment by jpountz** [comment]
I'm adding the `release highlight` label as I know many people including @qhoxie expressed interest for efficient recency and geo-distance boosting in the past.

**Comment by mayya-sharipova** [comment]
@cbuescher  thanks for such an extensive review - good feedback on the class design
@jpountz  thank for the review, Adrien. What you suggested makes sense, I will redesign this PR to have separate queries and only for  date and geo fields.

@jpountz  @cbuescher I am wondering what should be the name for these new queries. I have two options:
1) `distance_feature_date` and ` distance_feature_geo` - option in line with another similar query - `rank_feature` query
2) `recency_boosting` (for date) and `proximity_boosting` (for geo)  -  option to easier for users to understand what these queries are doing

**Comment by cbuescher** [comment]
> I wondered whether it would make more sense to have one query for all data types or one query per data type. 
> Having one query per data type might make things cleaner. For instance it's a pity that some validation only occurs at the shard level while we could do it at the coordinating node level

I agree that being able to validate at the coordinating node would be great. Separating the query however comes with some burden (two query builders, test, documentations etc...) that might be getting even bigger if we choose to add different proximity types later. Maybe it would be possible to stay with one `distance_feature` query but introduce a mandatory `type` argument (e.g. `date`, `geo` etc...) that we then could use for validation on the coordinating node? @jpountz wdyt?

**Comment by jpountz** [comment]
@cbuescher What would the API look like, are you thinking of something like below

```
"distance_feature": {
  "type": "geo",
  "field": "location",
  "origin": <origin>,
  "pivot": <pivot>,
  "boost" : <boost>
}
```

That said, we have a single query for the `range` query even though it suffers for the same issue, so maybe we should keep the current approach for consistency.

**Comment by cbuescher** [comment]
@jpountz yes, thats what I had in mind. This way I think we can check the input parameters for consistency already on the coordinating node when parsing the query, e.g. check that dates can be parsed etc...

**Comment by mayya-sharipova** [comment]
@jpountz @cbuescher  thanks for suggestions. I will go then with the format:
```
"distance_feature": {
  "type": "geo",
  "field": "location",
  "origin": <origin>,
  "pivot": <pivot>,
  "boost" : <boost>
}
```
For the license, it looks like we are sticking with OSS.

**Comment by jimczi** [comment]
> That said, we have a single query for the range query even though it suffers for the same issue, so maybe we should keep the current approach for consistency.

And for simplicity ;), exposing the `type` argument just to allow parsing on the coordinating node seems overkill to me and I am not sure that `type` would be enough to handle `date` format, ...
I am not too concerned by the fact that parsing errors will be detected at the shard level, most of our queries that require parsing have the same limitations so IMO we shouldn't add extra arguments that are useful only internally.

**Comment by cbuescher** [comment]
> I am not too concerned by the fact that parsing errors will be detected at the shard level, most of our queries that require parsing have the same limitations so IMO we shouldn't add extra arguments that are useful only internally.

I'm okay with either way that doesn't involve adding several differen query builders etc...

**Comment by jpountz** [comment]
Thanks @jimczi for chiming in, agreed.

File: server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: @cbuescher I have added the documentation for `parseFromString`. I have not added any checks for invalid input, as these checks are done withing interal functions that `parseFromString` calls.

**Comment by mayya-sharipova** [comment]
@cbuescher @jpountz  Thanks for your review. This is ready for another round of review. Changes made:

- Modify distance_feature query to be on date, date_nanos and geo ields only. Exclude long field
- Make `origin` as an inner class

**Comment by mayya-sharipova** [comment]
@elasticmachine elasticsearch-ci/1

**Comment by mayya-sharipova** [comment]
@elasticmachine  test this please

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: the cast from String might raise a ClassCastException?

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: our general policy is to be lenient when it comes to unmapped fields, to make cross-index search easier. I'd return a `MatchNoDocsQuery` instead.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: I don't like taking longs as-is, we should always go through `parseToLong` imo. I'd like to avoid that users need to be aware of the internal resolution of the field.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: can you add the type of the field to the error message?

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: can you add the class of the object to the error message?

**Review by jpountz — CHANGES_REQUESTED** [review]
I left some comments.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: nit: add a newline befor the constructor

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: can be package private I think

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: can be package private I think

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: can be package private I think

**Review by cbuescher — APPROVED** [review]
Left some more minor comments, but nothing huge that needs another review from my side I think.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: @jpountz  `parseToLong` does NOT work for date_nanos field when parsing a huge long value (an expected value for `date_nanos`). 
`DateFieldType::parseToLong`  uses `JavaDateMathParser` (instead of desired `JodaDateMathParser`) even if a  date type is `date_nanos`. Or should I pass a `forcedDateParser` to `parseToLong` depending on the fieldType's resolution?

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: @jpountz  Please disregard my previous comment here. I assumed that `date_nanos` except a long value as nanoseconds-since-the-epoch, but actually the long value should be  milliseconds-since-the-epoch.   
I have made the changes as you suggested.

**Comment by mayya-sharipova** [comment]
@cbuescher @jpountz  Thanks for another round of the  review.  I have addressed all your comments in the last commits.

**Comment by mayya-sharipova** [comment]
@elasticmachine  run elasticsearch-ci/packaging-sample

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: It's a bit confusing to pass a default value here given that the origin can't be null.

File: server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java]: let's reject nulls in above constructors

File: server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java]: Let's avoid using now to make failures reproducible?

File: server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java]: let's avoid using currentTimeMillis to make things reproducible?

**Review by jpountz — APPROVED** [review]
I left some minor comments. Other than that LGTM.

**Pull request merged**
